### PR TITLE
Add option to cancel when editing instructors #5608

### LINKED ISSUE #5608: Add option to cancel when editing instructors
Currently, at the "Edit Course Details" page, after clicking the "Edit" button to edit an instructor, there is no option to cancel, only options to "Delete" or to "Save changes".

This is counter-intuitive and inconsistent with the editing of feedback questions (which otherwise has similar UI).

### PR DESCRIPTION
**Pull request by JosephineKwa**
Fixes #5608 

Cancel button appears after clicking the edit button:
cancel

### CODE REVIEW

**Comment by chowyb** [comment]
With the "cancel" button, we might now need a way to change the options back to their original values as well upon cancelling. However, doing so in this PR may take a significant amount of time (ref #5468) especially when taking custom instructors into account.

Perhaps we could create a separate issue for this?

File: src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java [hunk 1]
```diff
@@ -119,6 +119,12 @@ private CourseEditInstructorPanel createInstructorPanel(InstructorAttributes cur
                                                        "javascript:;", Const.Tooltips.COURSE_INSTRUCTOR_EDIT,
                                                        null, isDisabled));
             
+            buttonContent = "<span class=\"glyphicon glyphicon-remove\"></span> Cancel";
+            instructorPanel.setCancelButton(createButton(buttonContent, "btn btn-primary btn-xs",
+                                                       "instrCancelLink" + instructorPanel.getIndex(),
+                                                       "javascript:;", Const.Tooltips.COURSE_INSTRUCTOR_CANCEL_EDIT,
+                                                       null, isDisabled));
```
Comment (chowyb) [inline src/main/java/teammates/ui/controller/InstructorCourseEditPageData.java]: Might need to shift these 3 lines another 2 spaces to the right

File: src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java [hunk 2]
```diff
@@ -121,6 +121,24 @@ public boolean clickEditInstructorLink(int instrNum) {
         return isEditable;
     }
     
+    public void clickCancelEditInstructorLink(int instrNum) {
+        getCancelEditInstructorLink(instrNum).click();
+    }
+    
+    public boolean isInstructorFormNotEditable(int instrNum) {
+        WebElement saveButton = getSaveInstructorButton(instrNum);
+        waitForElementToDisappear(By.id("btnSaveInstructor" + instrNum));
```
Comment (chowyb) [inline src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java]: Feels a bit odd having this here when the method call above uses the same selector, maybe we could extract the `By` object from `getSaveInstructorButton` as a separate method and use it here?

File: src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java [hunk 3]
```diff
@@ -121,6 +121,24 @@ public boolean clickEditInstructorLink(int instrNum) {
...
+        
+        WebElement editInstructorNameTextBox = getNameField(instrNum);
+        WebElement editInstructorEmailTextBox = getEmailField(instrNum);
+        
+        boolean isNotEditable = !editInstructorNameTextBox.isEnabled()
+                             && !editInstructorEmailTextBox.isEnabled()
+                             && !saveButton.isDisplayed();
```
Comment (chowyb) [inline src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java]: Shifting the `&&`s 3 spaces to the right to be in line with `!e` in the line above would be good :)

File: src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java [hunk 4]
```diff
@@ -121,6 +121,24 @@ public boolean clickEditInstructorLink(int instrNum) {
         return isEditable;
     }
     
+    public void clickCancelEditInstructorLink(int instrNum) {
+        getCancelEditInstructorLink(instrNum).click();
+    }
+    
+    public boolean isInstructorFormNotEditable(int instrNum) {
```
Comment (chowyb) [inline src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java]: Returning a `notEditable` boolean seems to go against the convention of booleans, but I guess this is due to implementation, as using `waitForElementVisibility` instead may result in false positives.

**Comment by chowyb** [comment]
@JosephineKwa some minor changes requested + clarification on reverting options :)

File: src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java (hunk 4 above)
Comment (JosephineKwa) [inline src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java]: `isInstructorFormEditable` would result in the boolean condition being
```
boolean isEditable = editInstructorNameTextBox.isEnabled()
                     && editInstructorEmailTextBox.isEnabled()
                     && saveButton.isDisplayed();
```
Which means that as long as one of the above elements is not enabled/displayed the form is considered as uneditable, which would also be a false positive.

Perhaps I could rename the method to `isInstructorFormUneditable` or `isInstructorEditFormDisabled` to avoid including `not` in the method name.

**Comment by JosephineKwa** [comment]
I will work on reverting options in this PR too

File: src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java (hunk 4 above)
Comment (chowyb) [inline src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java]: Going by pure boolean logic, the reverse of `isNotEditable` would be:

```
boolean isEditable = editInstructorNameTextBox.isEnabled()
                     || editInstructorEmailTextBox.isEnabled()
                     || saveButton.isEnabled();
```

which probably doesn't make too much sense either.

And yep renaming it to `isInstructorEditFormDisabled` would be good :)

File: src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java (hunk 2 above)
Comment (JosephineKwa) [inline src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java]: Perhaps I could omit checking whether the save button is displayed on line 137? Since there is already a wait for it to disappear. I will then be able to omit `WebElement saveButton = getSaveInstructorButton(instrNum);`

Comment (chowyb) [inline src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java]: Yep that could be possible.

On a related note, does having this line (the `waitForElementToDisappear`) mean that it times out (and throws a `TimeoutException`) if the save button is present, and this method is unlikely to properly return a `false` value?

Comment (JosephineKwa) [inline src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java]: Yup

**Comment by YongJieYongJie** [comment]
@JosephineKwa for reverting, maybe you can see my pr #5497, specifically [`instructorFeedbackEdit.js`](https://github.com/TEAMMATES/teammates/blob/10ae1bc9b40d961901f09bc751cf62b7a7e64798/src/main/webapp/js/instructorFeedbackEdit.js) at [`enableEdit()`](https://github.com/TEAMMATES/teammates/blob/10ae1bc9b40d961901f09bc751cf62b7a7e64798/src/main/webapp/js/instructorFeedbackEdit.js#L199)

File: src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java (hunk 2 above)
Comment (chowyb) [inline src/test/java/teammates/test/pageobjects/InstructorCourseEditPage.java]: Alright then :)

**Comment by JosephineKwa** [comment]
So far using `document.getElementById().reset();` seems to work too, except for selecting the access level radio

**Comment by chowyb** [comment]
@JosephineKwa the cancelling works better now, but the behaviour of hiding and showing panels could be improved - ideally an instructor should only see the panel if it contains privileges that are different from the privileges at a higher level (eg. section level privileges showing only if they are different from course level privileges), so that privileges can be cascaded properly.

While the `is___Special` variables used in JSTL to determine what is shown cannot be accessed directly in JS, you should be able to use the values of the `<input name="issectiongroup{sectionNum}set">` and `<input name="issectiongroup{sectionNum}sessionset">` elements (both hidden inputs) in the instructor panels to determine if the panel was shown initially or not :)

**Comment by JosephineKwa** [comment]
@chowyb could you tell me the steps to replicate this?

**Comment by chowyb** [comment]
@JosephineKwa as described earlier, but posting this here in case any other devs are interested:

1. Add a custom instructor with all privileges granted, but do not use "Give different permissions for a specific section". Save the instructor.
editins1
2. Try editing the newly created instructor. All section and session panels should now appear. Under the previous behaviour, none of them would be shown.
3. Editing any privileges at course level, followed by saving at this stage would result in the privileges not being cascaded downwards, while under the previous behaviour, it would cascade.

Everything else looks good though, so it should be good to merge after this is handled :)

**Comment by chowyb** [comment]
@JosephineKwa as mentioned, there is currently a bug in the code due to a change I introduced earlier (ref. #5272, #5723), where section panels were reordered, and therefore assuming the indices of the section permission panels are in the order (0, 1, 2...) is not valid.

An example of how this affects the current PR is as follows:

1. Add (or edit) a custom instructor, and click "Give different permissions for a specific section".
2. Select any section, but make sure the first section remains unchecked across all panels.
3. Select any permissions and save the change.
4. Try to edit the same instructor again. An additional section permission panel should be displayed with no sections checked.
Note: cancelling and editing this instructor again will result in the additional section not being displayed, which is expected behaviour.

I'll be working on fixing the ordering issue, but do you want to continue working on this PR in the meantime?

**Comment by chowyb** [comment]
@JosephineKwa I merged in the PR for the section panels, so merging master into this branch should fix the bug (I hope)

**Comment by JosephineKwa** [comment]
Ready for review

**Comment by chowyb** [comment]
Behaviour looks good now :+1:

File: src/main/webapp/WEB-INF/tags/instructor/course/courseEditTuneSectionPermissionsDiv.tag [hunk 5]
```diff
@@ -6,10 +6,10 @@
...
 <%@ attribute name="sectionRow" type="teammates.ui.template.CourseEditSectionRow" required="true" %>
 
-<div id="tuneSectionPermissionsDiv${panelIndex}ForInstructor${instructorIndex}" 
-        <c:if test="${not sectionRow.sectionSpecial}"> 
-            style="display: none;"
-        </c:if> >
+<div id="tuneSectionPermissionsDiv${panelIndex}ForInstructor${instructorIndex}" data-is-originally-displayed="${sectionRow.sectionSpecial}" 
+    <c:if test="${not sectionRow.sectionSpecial}"> 
+        style="display: none;"
+    </c:if>>
```
Comment (chowyb) [inline src/main/webapp/WEB-INF/tags/instructor/course/courseEditTuneSectionPermissionsDiv.tag]: Maybe write this as `</c:if> >` to be consistent with the previous style as well as the style currently used for `courseEditTuneSessionPermissionDiv.tag`

**Comment by chowyb** [comment]
Just one minor change requested, should be good to go after that :)

**Comment by chowyb** [comment]
Good to go! :)

**Comment by JosephineKwa** [comment]
@damithc pending your review on this PR too

File: src/test/java/teammates/test/cases/ui/browsertests/InstructorCourseEditPageUiTest.java [hunk 6]
```diff
@@ -587,6 +598,71 @@ private void testEditInstructorAction() throws Exception {
...
+        courseEditPage.clickSectionSelectionCheckBox(editInstructorIndex, 0, 0);
+        courseEditPage.clickSectionLevelPrivilegeLink(editInstructorIndex, 0, 0);
+        courseEditPage.clickSessionLevelInSectionLevel(editInstructorIndex, 0);
+        courseEditPage.clickAddSectionLevelPrivilegesLink(editInstructorIndex);
+        courseEditPage.clickSectionSelectionCheckBox(editInstructorIndex, 1, 1);
+        courseEditPage.clickSectionLevelPrivilegeLink(editInstructorIndex, 1, 0);
+        courseEditPage.clickAddSectionLevelPrivilegesLink(editInstructorIndex);
+        courseEditPage.clickSectionSelectionCheckBox(editInstructorIndex, 2, 2);
+        courseEditPage.clickSectionLevelPrivilegeLink(editInstructorIndex, 2, 0);
+        courseEditPage.clickCancelEditInstructorLink(editInstructorIndex);
```
Comment (damithc) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorCourseEditPageUiTest.java]: This code will be hard to follow for the reader, but I can't think of a better alternative.

Comment (chowyb) [inline src/test/java/teammates/test/cases/ui/browsertests/InstructorCourseEditPageUiTest.java]: The code here follows the style of other tests in the class, which has already been partially fixed in #5732; merging that branch in and changing the style here to match might help. 

Code such as `courseEditPage.clickSectionSelectionCheckBox(editInstructorIndex, 0, 0);` will remain as it is however, since there is no special meaning that is not derived directly from the numbers ("click the checkbox representing the first section in the first section group panel" (due to 0-based index) in this case).

Other more problematic code (testing feedback response comments, with a series of 4 integers as an identifier) comes to mind, but similarly to this case, there is no additional meaning that can be derived from the numbers.

**Comment by JosephineKwa** [comment]
I've replaced some of the magic numbers with constants, excluding numbers which represent ordered indices. @chowyb could you take a look?

**Comment by chowyb** [comment]
@JosephineKwa yep looks good :+1:
